### PR TITLE
Use `rasterio` package for grid masks

### LINKED ISSUE #23: Polygon clipping is still slow
Polygon clipping on large grids (BCCAQ scale... 10km over Canada) remains pretty slow (on the order of 5-10 seconds). This needs another round of optimization.

### PR DESCRIPTION
**Pull request by jameshiebert**
GDAL and [`rasterio`](https://github.com/mapbox/rasterio) provide methods for clipping a grid by some vector geometries (a.k.a. rasterizing or burning). GDAL is written in C++ and `rasterio` is written as Python C extension, and as such rasterizing is *really* fast. Using these packages, we're able to clip all of our test polygons in under 10ms.

The unequivocal speed advantage comes with the tradeoff of increased build complexity. GDAL can be notoriously difficult to build, and I've been unable to get Travis-CI to build GDAL's python package (or `rasterio`) against a system install of GDAL. `rasterio's` [Travis-CI config](https://github.com/mapbox/rasterio/blob/master/.travis.yml) [manually downloads the GDAL tarball](https://github.com/mapbox/rasterio/blob/master/scripts/travis_gdal_install.sh#L63) and builds it from source, so that's the approach that I've taken as well. This [adds about 20 lines](https://github.com/pacificclimate/climate-explorer-backend/compare/performance/rasterio?expand=1#diff-354f30a63fb0907d4ad57269548329e3R9) to our `.travis.yml` file and increases the build time to around 15 minutes.

IMO, the tradeoff is well worth it. It substantially decreases the amount of logic in `geo.py`, off-loads the work to a purpose-built library and provides a killer speed-up (6,000x?) that fixes #23 . But I'd like to request comments and see if everyone is OK with the increased build complexity.

### CODE REVIEW

File: ce/api/geo.py [hunk 1]
```diff
@@ -67,14 +68,9 @@ def __init__(self, func, maxsize=50):
...
-        nc, wkt = args
-        # If we have no key, automatic cache miss
-        if (not hasattr(nc, 'model_id')):
-            log.debug('Cache MISS (attribute \'model_id\' not found)')
-            return self.func(*args)
+        fname, wkt, varname = args
 
-        # Set key to model_id and wkt polygon
-        key = (nc.model_id, wkt)
+        key = (fname, wkt)
```
Comment (basilveerman) [inline ce/api/geo.py]: What's your motivation for this? We're going to have situations where it generates the same mask for the same grid multiple times again. We know that model_id isn't idea and we'd rather key on the grid_id as discussed, but this is a separate issue.

File: ce/api/geo.py [hunk 2]
```diff
@@ -104,55 +100,25 @@ def cache_clear(self):
...
+    dst_name = 'NETCDF:"{}":{}'.format(filename, variable)
+    raster = rasterio.open(dst_name, 'r', driver='NetCDF')
 
-    polygon_mask = pointsInPoly(lons, lats, poly)
+    if raster.affine == rasterio.Affine.identity():
+        raise Exception("Unable to determine projection parameters for GDAL "
+                        "dataset {}".format(dst_name))
 
-    return polygon_mask.mask
+    mask = rasterize((poly,), out_shape=raster.shape, transform=raster.affine, all_touched=True)
```
Comment (basilveerman) [inline ce/api/geo.py]: Nice how this entirely removes the issues of polygons being too small.

**Comment by basilveerman** [comment]
I agree. Amazing performance increases with code complexity decrease.

Travis taking that long for testing is a bit of a drag, but we should be running these tests ourselves anyway (aside: maybe we should set Travis-CI to only build pull requests rather than *every* commit). As for the increased build complexity, the new requirements were successfully built into a new Docker container without any issue. There may be additional complexity when manually creating a virtual environment for the back end, but this is minor compared to the gains we see.

**Comment by basilveerman** [comment]
Interesting however... querying the [old backend](http://docker1.pcic.uvic.ca:20003/api/timeseries?id_=tnxETCCDI_yr_BCCAQ-ANUSPLIN300-CanESM2_historical-rcp85_r1i1p1_1950-2100&variable=tnxETCCDI&area=) with no area selection takes ~2s. Hitting a [new version](http://docker1.pcic.uvic.ca:12000/api/timeseries?id_=tnxETCCDI_yr_BCCAQ-ANUSPLIN300-CanESM2_historical-rcp85_r1i1p1_1950-2100&variable=tnxETCCDI&area=) for the same dataset takes 22s.

Looking into it.

File: ce/api/geo.py (hunk 1 above)
Comment (jameshiebert) [inline ce/api/geo.py]: The netCDF4.Dataset object which *use to be* the parameter to `polygonToMask()` is no longer sufficient. `polygonToMask()` now has to open the file with both netCDF4 (to test for 0-360 grids) and `rasterio` to do the actual work.

File: ce/api/geo.py (hunk 2 above)
Comment (jameshiebert) [inline ce/api/geo.py]: Yeah, pretty nice. It still doesn't yet give us the RAT feature of weighting grid cells by percentage of each cell included, but I think that we can get that eventually (and our old code didn't give us that either).

**Comment by basilveerman** [comment]
@jameshiebert Take a look at this [branch](https://github.com/pacificclimate/climate-explorer-backend/tree/performance/rasterio-speedtest) I added some debug statements to.

| branch | no poly speed | no poly time (1/2 domain) | large poly speed | large poly time | small poly (metro van) speed | small poly time |
|---|---|---|---|---|---|---|
| master | 250-300 MB/s | ~1.6s | 175-200 MB/s | 25-30s | 175-200 MB/s | 25-30s | 
| performance/rasterio | 10-15 MB/s | ~25s | 175 - 200 MB/s | ~3s | 175 - 200 MB/s | ~3s |

The speeds should be representative of read speed since they're all for *after* polygon calculation, while the total time takes polygon calc into account.

Considering nothing substantial changed in the code path without a polygon, this is really weird. Can you replicate?

**Comment by jameshiebert** [comment]
So, I've investigated the performance of get_array and I don't think that the problem is necessarily there. I [added that file](https://github.com/pacificclimate/climate-explorer-backend/commit/e0cd3d03d6e7a38a8f5bd8322b3930edb454a6ad) to the test suite and included a performance test in the test suite that limits the response time to 30ms and all of the polygons (and no polygon) pass.

```
(env)james@basalt ~/code/git/climate-explorer-backend $ py.test -v -k get_array -s 2>/dev/null
============================================================== test session starts ===============================================================
platform linux -- Python 3.4.3, pytest-2.9.1, py-1.4.31, pluggy-0.3.1 -- /home/james/code/git/climate-explorer-backend/env/bin/python3.4
cachedir: .cache
rootdir: /home/james/code/git/climate-explorer-backend, inifile: pytest.ini
collected 83 items 

ce/tests/test_util.py::test_get_array[haida_gwaii_6-cgcm.nc-tasmax] 0.013779401779174805
PASSED
ce/tests/test_util.py::test_get_array[haida_gwaii_6-cgcm-tmin.nc-tasmin] 0.003861665725708008
PASSED
ce/tests/test_util.py::test_get_array[haida_gwaii_6-prism_pr_small.nc-pr] 0.007082462310791016
PASSED
ce/tests/test_util.py::test_get_array[haida_gwaii_6-bccaq_tnx.nc-tnxETCCDI] 0.027102947235107422
PASSED
ce/tests/test_util.py::test_get_array[bc_18-cgcm.nc-tasmax] 0.004057168960571289
PASSED
ce/tests/test_util.py::test_get_array[bc_18-cgcm-tmin.nc-tasmin] 0.0037920475006103516
PASSED
ce/tests/test_util.py::test_get_array[bc_18-prism_pr_small.nc-pr] 0.00672149658203125
PASSED
ce/tests/test_util.py::test_get_array[bc_18-bccaq_tnx.nc-tnxETCCDI] 0.024984359741210938
PASSED
ce/tests/test_util.py::test_get_array[fraser_plateau_270-cgcm.nc-tasmax] 0.007082462310791016
PASSED
ce/tests/test_util.py::test_get_array[fraser_plateau_270-cgcm-tmin.nc-tasmin] 0.007256269454956055
PASSED
ce/tests/test_util.py::test_get_array[fraser_plateau_270-prism_pr_small.nc-pr] 0.007554292678833008
PASSED
ce/tests/test_util.py::test_get_array[fraser_plateau_270-bccaq_tnx.nc-tnxETCCDI] 0.026453733444213867
PASSED
ce/tests/test_util.py::test_get_array[metro_van_10-cgcm.nc-tasmax] 0.004493236541748047
PASSED
ce/tests/test_util.py::test_get_array[metro_van_10-cgcm-tmin.nc-tasmin] 0.003854990005493164
PASSED
ce/tests/test_util.py::test_get_array[metro_van_10-prism_pr_small.nc-pr] 0.006267070770263672
PASSED
ce/tests/test_util.py::test_get_array[metro_van_10-bccaq_tnx.nc-tnxETCCDI] 0.02552938461303711
PASSED
ce/tests/test_util.py::test_get_array[no_polygon_0-cgcm.nc-tasmax] 0.0008177757263183594
PASSED
ce/tests/test_util.py::test_get_array[no_polygon_0-cgcm-tmin.nc-tasmin] 0.0007863044738769531
PASSED
ce/tests/test_util.py::test_get_array[no_polygon_0-prism_pr_small.nc-pr] 0.0016832351684570312
PASSED
ce/tests/test_util.py::test_get_array[no_polygon_0-bccaq_tnx.nc-tnxETCCDI] 0.008713006973266602
PASSED

====================================================== 63 tests deselected by '-kget_array' ======================================================
==================================================== 20 passed, 63 deselected in 1.35 seconds ====================================================
```
I didn't actually check in the testing file to the repo because it's like half a gig, so if you want to reproduce, you'll have to grab the file from pcic-storage yourself.

I'm going to look farther up the stack.The timeseries API call *does* make repeated calls to `get_array()`, and the file handling is still a little janky (i.e. the NetCDF file gets opened/maybe-closed on each call). Repeatedly opening a FUSE-mounted file has a high potential to be the culprit, but I'll look into it a bit more.

**Comment by jameshiebert** [comment]
I've moved up the stack and written a [performance test for the timeseries API call](https://github.com/pacificclimate/climate-explorer-backend/commit/5d626dd404afd8b8b904a5e70c0f197127e0f96d#diff-aa545f08105c1dab446bcbe534e4959fR181) and still cannot reproduce. The call using a polygon come in at around 2.5 seconds and the empty polygon calls come it at a little under 2 seconds. This certainly isn't fast and should still be improved. But it's nowhere near the 25 seconds that you're seeing. Could you run a profiler on one of these testing branches and get me the profiler output? Maybe that will shed more light on the issue.

```
(env)james@basalt ~/code/git/climate-explorer-backend $ py.test -v -k timeseries -s 2>/dev/null
============================================================== test session starts ===============================================================
platform linux -- Python 3.4.3, pytest-2.9.1, py-1.4.31, pluggy-0.3.1 -- /home/james/code/git/climate-explorer-backend/env/bin/python3.4
cachedir: .cache
rootdir: /home/james/code/git/climate-explorer-backend, inifile: pytest.ini
collected 88 items 

ce/tests/test_api.py::test_api_endpoints_are_callable[timeseries-query_params2] PASSED
ce/tests/test_api.py::test_timeseries[no_polygon_0] PASSED
ce/tests/test_api.py::test_timeseries[metro_van_10] PASSED
ce/tests/test_api.py::test_timeseries[bc_18] PASSED
ce/tests/test_api.py::test_timeseries[fraser_plateau_270] PASSED
ce/tests/test_api.py::test_timeseries[haida_gwaii_6] PASSED
ce/tests/test_api.py::test_timeseries_bad_id[None] PASSED
ce/tests/test_api.py::test_timeseries_bad_id[] PASSED
ce/tests/test_api.py::test_timeseries_bad_id[does-not-exist] PASSED
ce/tests/test_api.py::test_timeseries_speed[no_polygon_0] 1.7210447788238525
PASSED
ce/tests/test_api.py::test_timeseries_speed[metro_van_10] 2.613750696182251
PASSED
ce/tests/test_api.py::test_timeseries_speed[bc_18] 2.6565537452697754
PASSED
ce/tests/test_api.py::test_timeseries_speed[fraser_plateau_270] 2.629056215286255
PASSED
ce/tests/test_api.py::test_timeseries_speed[haida_gwaii_6] 2.646106243133545
PASSED

===================================================== 74 tests deselected by '-ktimeseries' ======================================================
=================================================== 14 passed, 74 deselected in 14.85 seconds ====================================================
```

**Comment by basilveerman** [comment]
I set up another backend on a compute node running Flask in a virtual environment:

```bash
time wget -qO- "http://puma.compute.pcic.uvic.ca:8910/api/timeseries?id_=tx10pETCCDI_yr_BCCAQ-ANUSPLIN300-CanESM2_historical-rcp85_r1i1p1_1950-2100&variable=tx10pETCCDI&area=POLYGON+((-123.82812509313226+52.02189177274704%2C+-123.82812509313226+54.46329802274705%2C+-121.19140634313227+54.46329802274705%2C+-121.19140634313227+52.02189177274704%2C+-123.82812509313226+52.02189177274704))" &> /dev/null
> real	0m2.350s

time wget -qO- "http://puma.compute.pcic.uvic.ca:8910/api/timeseries?id_=tx10pETCCDI_yr_BCCAQ-ANUSPLIN300-CanESM2_historical-rcp85_r1i1p1_1950-2100&variable=tx10pETCCDI&area=" &> /dev/null
> real	0m1.840s
```

Ok, that's what it should be.

Tried rebuilding Docker container:

```bash
git checkout performance/rasterio
docker build -t pcic/climate-explorer-backend:performance-rasterio .
sudo docker run --rm -it -p 8910:8000 -e "MDDB_DSN=postgresql://xxxxxxxx" -v /storage/data/scratch/bveerman/climatologies:/storage/data/scratch/bveerman/climatologies -v /storage/data/climate/CLIMDEX/CMIP5/multidecadal_means:/storage/data/climate/CLIMDEX/CMIP5/multidecadal_means -v /storage/data/climate/downscale/CMIP5/BCCAQ/climdex:/storage/data/climate/downscale/CMIP5/BCCAQ/climdex --name backend-rasterio pcic/climate-explorer-backend:performance-rasterio 
```

```bash
time wget -qO- "http://docker1.pcic.uvic.ca:8910/api/timeseries?id_=tx10pETCCDI_yr_BCCAQ-ANUSPLIN300-CanESM2_historical-rcp85_r1i1p1_1950-2100&variable=tx10pETCCDI&area=" &> /dev/null
> real	0m24.584s

time wget -qO- "http://docker1.pcic.uvic.ca:8910/api/timeseries?id_=fdETCCDI_yr_BCCAQ-ANUSPLIN300-CanESM2_historical-rcp85_r1i1p1_1950-2100&variable=fdETCCDI&area=POLYGON+((-124.21875009313227+51.33829802274704%2C+-124.21875009313227+53.68204802274704%2C+-122.16796884313226+53.68204802274704%2C+-122.16796884313226+51.33829802274704%2C+-124.21875009313227+51.33829802274704))" &> /dev/null
> real	0m2.730s
```

We see the performance issue again.

I'm rebuilding the base image (pcic/geospatial-python) to eliminate any possibility of older packages with bugs. If the issues persists, we can chalk this one up to Docker.

**Comment by basilveerman** [comment]
@jameshiebert: Got it.

In venv with numpy 1.11.0

```
name                                  ncall  tsub      ttot      tavg   
..ce/api/timeseries.py:67 <listcomp>  1      0.004967  2.493146  2.493146
/app/ce/api/util.py:42 get_array      151    0.794631  1.530049  0.010133
..umpy/core/fromnumeric.py:2632 mean  151    0.001283  0.950712  0.006296
..y/ma/core.py:4684 MaskedArray.mean  151    0.006056  0.949430  0.006288
..mpy/core/fromnumeric.py:347 repeat  151    0.000503  0.532952  0.003529
..py/ma/core.py:4437 MaskedArray.sum  151    0.028863  0.489085  0.003239
../ma/core.py:3931 MaskedArray.count  151    0.057590  0.454289  0.003009
..ma/core.py:3400 MaskedArray.filled  151    0.002855  0.340213  0.002253
```
In docker with numpy 1.8.2

```
name                                  ncall  tsub      ttot      tavg   
..ce/api/timeseries.py:67 <listcomp>  2      0.039845  23.92250  11.96125
/app/ce/api/util.py:42 get_array      302    1.678396  22.10160  0.073184
..ages/numpy/ma/core.py:2624 __new__  604    0.066985  19.63970  0.032516
..py/core/fromnumeric.py:1011 resize  151    0.816621  19.53177  0.129349
..umpy/core/fromnumeric.py:2632 mean  302    0.002601  1.765478  0.005846
..y/ma/core.py:4684 MaskedArray.mean  302    0.011630  1.762877  0.005837
..py/ma/core.py:4437 MaskedArray.sum  302    0.047887  0.883850  0.002927
../ma/core.py:3931 MaskedArray.count  302    0.107718  0.867397  0.002872
..ma/core.py:3400 MaskedArray.filled  302    0.005418  0.598756  0.001983
..mpy/core/fromnumeric.py:347 repeat  151    0.000503  0.532952  0.003529
..ges/numpy/core/_methods.py:23 _sum  604    0.003508  0.363065  0.000601
```
In docker with numpy 1.11.0
```
name                                  ncall  tsub      ttot      tavg      
..ce/api/timeseries.py:67 <listcomp>  1      0.003778  1.338823  1.338823
/app/ce/api/util.py:42 get_array      151    0.719849  0.864006  0.005722
..umpy/core/fromnumeric.py:2796 mean  151    0.000916  0.464060  0.003073
..y/ma/core.py:5014 MaskedArray.mean  151    0.003605  0.463144  0.003067
..py/ma/core.py:4770 MaskedArray.sum  151    0.025267  0.359799  0.002383
..ma/core.py:3576 MaskedArray.filled  151    0.002916  0.280693  0.001859
..ges/numpy/core/_methods.py:31 _sum  302    0.000693  0.145002  0.000480
../ma/core.py:4224 MaskedArray.count  151    0.001982  0.099740  0.000661
```

Unfortunately, numpy 1.8.2 is the latest version in Ubuntu packages. I'm rebuilding base Docker images to include the updated version on PyPI.

Still haven't a clue why this cropped up in this PR when the same execution path (essentially) didn't have the issue in master.

**Comment by basilveerman** [comment]
@mdfischer, do you have any comments regarding Jame's initial questions?

**Comment by jameshiebert** [comment]
With this set of changes, the netcdf file *and* the filename get passed around as arguments to a couple functions. It's a bit awkward and I don't really like it, but I can't see a better way without the `netCDF4.Dataset.filepath()` method working properly (which it doesn't).

**Comment by mdfischer** [comment]
Sorry, just got to reading this thread now.  Very interesting that this issue didn't show up before.  Regarding James' initial questions, I'm OK with the increased build time of GDAL tradeoff for performance gains, and agree that it would be much better if TravisCI only builds on PRs to master rather than on every commit.

**Pull request merged**